### ISSUE TITLE
Is there examples or documentation?

### ISSUE THREAD

**Issue by haha150** [nbgallery/nbgallery#52]
Hi,

So i am trying to extend the framework by for example changing some of the frontend/backend code, however it is not clear to me how the flow is executed. Do you have any documentation or examples of how it is interacting? For example, when you click on the login button and fill in email/password and click on submit, where/what is this submit form invoking in the codebase/backend (including mysql invocation)?

Also i would like to know how i can, from adding an extention, to invoking my extention from code either frontend or backend.

**Comment by jeffyjefflabs** [comment]
The extension system could definitely be better documented.  There is an overview [here](https://github.com/nbgallery/nbgallery/blob/master/docs/extensions.md) and a few examples [here](https://github.com/nbgallery/nbgallery/tree/master/samples).  I've been meaning to add some more stubs there as examples.

User authentication is handled by the [Devise gem](https://github.com/plataformatec/devise) and is hard to follow - I would suggest reading about Devise first.  One of the sample extensions at the link above is a stub for how you would plug in a custom Devise authentication strategy.

What sort of extension are you thinking about?

**Comment by haha150** [comment]
Basically, i am trying to invoke my external authentication extension which i have implemented in ruby usinb LDAP. I just dont know how to invoke the extension at the moment a user clicks on the "login" button.

**Comment by jeffyjefflabs** [comment]
I would see if that's possible using Devise - they have an example [on their wiki](https://github.com/plataformatec/devise/wiki/How-To:-Authenticate-via-LDAP) and [this gem](https://github.com/cschiewek/devise_ldap_authenticatable) looks promising as well.

On the nbgallery side, the "Devise boundary" is `set_user` in the [application_controller](https://github.com/nbgallery/nbgallery/blob/master/app/controllers/application_controller.rb#L50), which gets called at the beginning of every request.  The `user_signed_in?` and `current_user` functions referenced there are part of the Devise framework.

**Comment by haha150** [comment]
So, i followed the example from Devise and added my implementation, so it is not an extension basically, but part of the code base.

Another thing i wonder is, how can i modify the login form so that it does not require valid email addresses? wants "@" in the email field where as i dont want to enforce this.

**Comment by jeffyjefflabs** [comment]
The back-end will also validate that the email address contains "@".  Note that user accounts have an email address (never shown in the UI) and a user_name (used in the UI) but login is by email.

Users do need to be in the local nbgallery database, so yes you need to figure out a way to create the objects.  We do that from our extension - after validating with the external service, if they don't exist in the nbgallery database we create a new user object (directly setting fields like confirmed_at and approved), and then log them in.  So you might still need an extension that works with the LDAP Devise plugin somehow.

**Comment by mcrutch** [comment]
Specifically, this is how it validates email - https://github.com/nbgallery/nbgallery/blob/master/lib/gallery_lib.rb#L10 which precludes any @hostname type email addresses if you are testing things on a local box/network.  It requires @domain.tld style addresses.  (Was a gotcha for me while creating some dummy accounts for internal testing)

**Comment by manfromjupyter** [comment]
Closing as resolved.
